### ISSUE TITLE
time cost in inference mode

### ISSUE THREAD

**Issue by JackLee396** [pierluigiferrari/ssd_keras#61]
Thank you for sharing the code! I trained ssd300 on my own datasets and made detection on a single 1024x1024 image with batchsize=1, and it took about 1.8s on GPU NVIDIA 1080Ti , which is much more than the time mentioned in the paper. And I also made detection on CPU, and it only took a bit longer which is 3.8s.  I have no idea why that happened... Anybody who can explained about that ? Thank you!

**Comment by pierluigiferrari** [comment]
I've just ran 100 predictions with an SSD300 on Pascal VOC (i.e. 20 classes) at batch size 1 on a GTX1070 and here are the results:

- Raw prediction (i.e. `model.predict()`): 0.0227 seconds on average, or 44.05 FPS
- Decoding (i.e. `decode_y()`): 0.0104 seconds on average

This makes a total of 0.0331 seconds per prediction, or 30.21 FPS.

Note that `decode_y()` runs on the CPU, which is not ideal - it adds almost 50% of the time of a forward pass through the Keras model. One item on the ToDo list in the README is to write a decoding layer to move the decoding process from Numpy into TensorFlow, but I haven't gotten to doing that yet.

As for why your predictions are so slow:

I can only partially explain that. For starters, note that the image size is key for the prediction time. The high speed of 46 FPS stated for the original Caffe models is for SSD300, i.e. 300x300 input size. SSD512 is already a lot slower.

Your 1024x1024 input makes almost 12 times as many pixels per forward pass as SSD300. I can't say how exactly that translates into compute time, but it sure is significant.

1.8 seconds per prediction is an eternity though, that's 54-80 times slower than what I was measuring in my tests above (and my GPU is even weaker than yours), which surely cannot only be explained by the high image resolution you're using. I don't know what the answer is, I can only tell you that I can't confirm these long prediction times.

Are the number of classes in your model very large? Are you 100% sure you're actually using your GPU and the GPU has enough free memory?

**Comment by JackLee396** [comment]
Thank you for your reply! I've found the possible reasons. 
It seems like the GPU need a warm-up before forwarding the network.
 When I increase the batchsize and measure the time again, it turn out that only the first inference take a longer time than the followings, and the average time of the inference except the first one is 80ms, which is much more close to the time in the paper.

**Comment by pierluigiferrari** [comment]
Yeah, that sounds more plausible for your image size. Great!
